### PR TITLE
Fetch projects using the paginated list endpoint

### PR DESCRIPTION
**Pull request by erpas**
Also fixed multiple project managers creation issue.

### CODE REVIEW

**Comment by erpas** [comment]
Currently, Mergin plugin only allows to fetch 100 projects in a group.

This PR allows for getting more projects in chunks of 50. Each projects group has the Fetch more option in the right-click menu, see below. Also the numbers of projects already downloaded and available on the server are shown.

obraz

Additionally, a "Double-click for more..." item is added at the end of the list, if there are more projects to fetch from the service.

obraz

File: Mergin/plugin.py [hunk 1]
```diff
@@ -162,61 +170,87 @@ def add_action(
...
         """Open plugin configuration dialog."""
-        if self.data_item_provider.root_item is None:
-            QMessageBox.warning(
-                None,
-                "Mergin Plugin",
-                "Mergin plugin first use requires QGIS restart.",
-            )
-            return
-        self.data_item_provider.root_item.configure()
+        self.show_browser_panel()
```
Comment (wonder-sk) [inline Mergin/plugin.py:226]: :thinking:  not sure if we need this... and especially to do it before opening the config dialog... I would at least move that further down when the config dialog was accepted...

File: Mergin/plugin.py [hunk 2]
```diff
@@ -262,13 +294,14 @@ def unload(self):
...
 
-    def __init__(self, parent, project, mc, project_manager):
+    def __init__(self, parent, project, project_manager):
         self.project = project
         self.project_name = posixpath.join(
             project["namespace"], project["name"]
         )  # we need posix path for server API calls
         QgsDataItem.__init__(self, QgsDataItem.Collection, parent, self.project_name, "/Mergin/" + self.project_name)
         self.path = mergin_project_local_path(self.project_name)
+        self.setSortKey(f"1 {self.name()}")
```
Comment (wonder-sk) [inline Mergin/plugin.py:304]: This would deserve probably a comment :-) not sure why it is there...

File: Mergin/plugin.py [hunk 3]
```diff
@@ -575,71 +625,116 @@ def __init__(self, parent, grp_name, grp_filter, icon, order, project_manager):
...
+                self.removeChildItem(self.fetch_more_item)
+            except RuntimeError:
+                pass
+            self.fetch_more_item = None
+        fetched_count = self.rowCount() + len(self.projects)
+        if fetched_count < self.total_projects_count:
+            self.fetch_more_item = FetchMoreItem(self)
+            self.fetch_more_item.setState(QgsDataItem.Populated)
+            sip.transferto(self.fetch_more_item, self)
+        group_name = f"{self.group_name} ({fetched_count}/{self.total_projects_count})"
```
Comment (wonder-sk) [inline Mergin/plugin.py:683]: I don't really like the idea to show "XXX (fetched/total)" for each group, sorry :-)  maybe just show "XXX (total)" but including the number of fetched projects is IMHO not needed...

File: Mergin/plugin.py [hunk 4]
```diff
@@ -575,71 +625,116 @@ def __init__(self, parent, grp_name, grp_filter, icon, order, project_manager):
...
+        self.setName(group_name)
+
+    def fetch_more(self):
+        """Fetch another page of projects and add them to the group item."""
+        if self.total_projects_count - self.rowCount() <= 0:
+            QMessageBox.information(None, "Fetch Mergin Projects", "All projects already listed.")
```
Comment (wonder-sk) [inline Mergin/plugin.py:689]: Can this ever happen? Do we allow people to fetch more if there's nothing else to fetch?

File: Mergin/plugin.py [hunk 5]
```diff
@@ -575,71 +625,116 @@ def __init__(self, parent, grp_name, grp_filter, icon, order, project_manager):
...
         except URLError:
             error_item = QgsErrorItem(self, "Failed to get projects from server", "/Mergin/error")
             sip.transferto(error_item, self)
             return [error_item]
         except Exception as err:
             error_item = QgsErrorItem(self, "Error: {}".format(str(err)), "/Mergin/error")
             sip.transferto(error_item, self)
-            return [error_item]
+            return None, [error_item]
+        return None
```
Comment (wonder-sk) [inline Mergin/plugin.py:652]: Hmm what happens with the return values? in different occasions we return `[error_item]` or `None, [error_item]` or `None` ? What are we supposed to return?

File: Mergin/plugin.py (hunk 2 above)
Comment (erpas) [reply Mergin/plugin.py:304]: It is here to:
* make sure projects are listed first ("1") before the `Fetch more` item
* get the projects ordered by namespace AND name - currently we can only ordered by a single project attr on the server - that was namespace to make fetching more project process consistent. 

When the `order_params` of the `paginated_projects_list` gets into play on the server and we will get the list already sorted, but still we need a way to put the `Fetch more...` item at the bottom of a group tree branch.

File: Mergin/plugin.py (hunk 3 above)
Comment (erpas) [reply Mergin/plugin.py:683]: I was think about users fetching more of Explore group projects, so they clearly know how many of them are already fetched - there are nearly 1K of them ATM. It is true that users have the projects ordered alphabetically and can kind of guess that there is plenty of them to be fetched if they see D as the first letter of projects name for example.
It is also true, that there won't be Fetch more... item at the bottom of the list if there is nothing more to get.
And also true that for "My Projects" and "Shared with me" projects there not that many of them and it looks a bit odd.

I will show only the total number of projects in each group, as you suggest.
How about having the complete count info in the tool tip for group items?

File: Mergin/plugin.py (hunk 4 above)
Comment (erpas) [reply Mergin/plugin.py:689]: Nope, this is a leftover from an earlier stage when the "Fetch more..." item was not removed from the list when all projects were already fetched.
Fixed.

File: Mergin/plugin.py [hunk 6]
```diff
@@ -59,7 +62,10 @@ def __init__(self, iface):
         self.actions_always_on = []
         self.menu = "Mergin Plugin"
         self.mergin_proj_dir = None
+        self.mc = None
         self.manager = None
+        self.wizard = None
+        self.error = None
```
[the remark this answers is not in the document]
Comment (erpas) [reply Mergin/plugin.py:68]: Yep.

File: Mergin/plugin.py [hunk 7]
```diff
@@ -59,7 +62,10 @@ def __init__(self, iface):
         self.actions_always_on = []
         self.menu = "Mergin Plugin"
         self.mergin_proj_dir = None
+        self.mc = None
         self.manager = None
+        self.wizard = None
```
[the remark this answers is not in the document]
Comment (erpas) [reply Mergin/plugin.py:67]: Yep.

File: Mergin/plugin.py (hunk 1 above)
Comment (erpas) [reply Mergin/plugin.py:226]: Moved down.

File: Mergin/plugin.py (hunk 5 above)
Comment (erpas) [reply Mergin/plugin.py:652]: Yep, this is a mess. We either need None (all OK) of an error item (as a list item), if something goes wrong.

**Review by wonder-sk — APPROVED** [review]

**Pull request merged**